### PR TITLE
[RLlib][DO NOT MERGE] PPO RLHF Example

### PR DESCRIPTION
**Pull request by kouroshHakha**
## Why are these changes needed?

This PR is an attempt to create an RLHF pipeline example in rllib. Doing this will achieve a couple of things:

1. Evaluate if our RLModule / Learner APIs are simple to use, robust and sufficient for this application. 
2. It will put a spotlight on other short-comings of RLlib that needs to be fixed (sampler issues)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/examples/rlhf/ppo_ft/rlhf_env.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+            probs_list.append(probs)
+
+            # Update the input_ids with the generated token
+            model_in = torch.cat([model_in, next_token], dim=-1)
+
+            if next_token.item() == eos_token_id:
+                break
+
+        # Decode and print the generated sequence
+        generated_tokens = torch.cat(generated_sequence, dim=-1)
```
Comment (gjoliver) [inline rllib/examples/rlhf/ppo_ft/rlhf_env.py:49]: if we are doing sequences in 1 shot, how are we gonna take discounted future return into consideration?
should the reward function score every single token generated?

Comment (gjoliver) [reply rllib/examples/rlhf/ppo_ft/rlhf_env.py:49]: I guess to clarify my confusion, for example, the kind of behavior we are trying to encourage is when we are in the middle of a sentence, the chosen token should allow us to conclude the whole sentence sooner.
so the training sample should only be the partial sentence, and the next word choice?
I don't quite get how scoring the whole sequence would help.

**Review by gjoliver — COMMENTED** [review]
I agree at a high level, this is what we need.
thanks for starting this.

File: rllib/examples/rlhf/ppo_ft/notes.txt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+- _initialize_loss_from_dummy_batch() is a stupid logic that we should get rid off experimentally. I have added an experimental flag to disable it upon request. We should later get rid of this thing.
+
+
+04/06
+
+- I got the RLHF env working. 
+- The tricky part is getting observation / action spaces to correspond to what is expected. There is not much value in this except that when using RLlib obs and action space are the things that gets passed to the model constructor (which is ignored in case of LLMs)
+- Another tricky that required some thinking was formulating the env as bandit in LLMs.
+You want the env to read the dataset and accept a completion and then call the episode done. This type of env can be massively batched and should run on gpus, and we should not necessarily adhere to RLlib's assumption on sequential decision making envs, and also keeping things in numpy land.
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/examples/rlhf/ppo_ft/notes.txt:20]: While these models are big, they can totally live behind some end-point that you query. The network cost can be insignificant compared to the model living the same machine as your env. Think of doing batch inference to an end-point that has gpt-3 behind it.

File: rllib/examples/rlhf/ppo_ft/rlhf_env.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+        # This is the base tokenizer
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(
+            config["tokenizer_path"]
+        )
+
+        # This is the SFT model, used for computing the kl-divergence penalties.
+        self.sft_model = transformers.GPT2LMHeadModel.from_pretrained(
+            config["sft_model_path"]
+        )
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/examples/rlhf/ppo_ft/rlhf_env.py:122]: yep, it's pre-trained via sft not via unsupervised language modeling objective. It's just a reminder on which model it is.

File: rllib/examples/rlhf/ppo_ft/rlhf_env.py (hunk 1 above)
Comment (kouroshHakha) [reply rllib/examples/rlhf/ppo_ft/rlhf_env.py:49]: This is how instructGPT introduces RLHF. the truth is that you don't get any evaluation signal until the end of the sentence. i.e. the reward model is queried only when the answer is complete to evaluate whether the response was good for the input instruction. So your discount factor should be 1 here because you want to propagate that final reward equally to every token decision you have made so far. Also this way you'll only do one forward pass on RM and b/c you don't need to call it auto-regressively. However if it's a conversation things differ, the model can ask for clarifications etc. but in conversation you also have to generate full sentences on each turn. If you incorporate RLHF in conversation fashion, then your model can learn to ask you clarifying questions, because that will help it give more satisfying answers faster.

File: rllib/examples/rlhf/ppo_ft/notes.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/examples/rlhf/ppo_ft/notes.txt:20]: I see. interesting 😄

File: rllib/examples/rlhf/ppo_ft/rlhf_env.py (hunk 1 above)
Comment (gjoliver) [reply rllib/examples/rlhf/ppo_ft/rlhf_env.py:49]: oh. I see.

File: rllib/examples/rlhf/ppo_ft/rlhf_env.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        sequence = action["sequence"]
+        response_mask = action["response_mask"]
+        attention_mask = action["attention_mask"]
+        logits = action["logits"]
+
+        n_response_tokens = response_mask.sum()
+
+        r_align = self.reward_model(sequence, attention_mask, response_mask)
+        r_align = r_align.item()
+
```
Comment (zhangjian94cn) [inline rllib/examples/rlhf/ppo_ft/rlhf_env.py:194]: Maybe you can add a configurable discount to the reward, as done [here](https://github.com/nebuly-ai/nebullvm/blob/013084714b7f26ca41afc4662df65432689abcfe/apps/accelerate/chatllama/chatllama/rlhf/trainer.py#L962)

```python
gamma = self.config.trainer.gamma_discounted
discounted_rewards = torch.zeros_like(old_values)
for i in range(discounted_rewards.shape[1]):
    for j in range(i, discounted_rewards.shape[1]):
        discounted_rewards[:, i] += (
            gamma ** (j - i) * rewards[:, j]
        )
```

**Comment by PurpleSand123** [comment]
Hello, could you please let me know the reason behind the closure of this PR?
I'm planning to apply RLHF with Ray in my project and would appreciate any feedback on why the example didn't merge. Thank you for your help.

**Comment by kouroshHakha** [comment]
This PR was more of a prototype at the time and probably won't scale to super large models. Also RLHF does not have much demand in the market hence closing this PR.

**Comment by PurpleSand123** [comment]
Thank you for your response. Do you mean RLHF can function properly in Ray RLlib?
